### PR TITLE
Issue #3174544 and #3174547 by Kingdutch: Implement basic GraphQL support for users and profile data

### PR DESCRIPTION
**Pull request by Kingdutch**
## Problem
Open Social is slowly moving towards an API first approach.

This issue serves as a meta issue for community discussion and to track to work that will be required to build the new APIs as well as what can be done in the existing themes to utilise the new API structure.

## Solution
Open Social has selected version 4 of <a href="https://www.drupal.org/project/graphql">the GraphQL module</a> as it allows us to precisely control the shape of the Graph that we want to present. This allows us to better decouple a clear data graph from the internal data representation.

This pull request seeks to implement the base GraphQL module as well as the initial exposure of profile and user data. No authentication for external applications is implemented yet. The current implementation requires being authenticated with the default Drupal browser session.

The included GraphQL module is hidden and experimental for Open Social's internal use and evaluation at the moment and no support is yet provided until the covered parts of the Graph are deemed stable.

## Issue tracker
- [Issue 3174544: [GraphQL] Implement base GraphQL module](https://www.drupal.org/project/social/issues/3174544)
- [Issue 3174547: [GraphQL] Implement basic user and profile querying in the Graph](https://www.drupal.org/project/social/issues/3174547)
- [Issue 3191701: User creation breaks if no profile type exists](https://www.drupal.org/project/social/issues/3191701)
- [Issue 3191702: Provided user schema is incorrect and unused](https://www.drupal.org/project/social/issues/3191702)
- [Issue 3191703: Allow attaching libraries to AccountHeader elements](https://www.drupal.org/project/social/issues/3191703)
- [Issue 3191704: Open Social should only run its own PHPUnit tests](https://www.drupal.org/project/social/issues/3191704)

## How to test
- [x] Enable the social_grahpql module
- [x] As a LU go to `/admin/config/graphql/servers/manage/open_social_graphql/explorer`
- [x] You should be able to query user and profile data.

## Documentation
A walk-through of how this all works is available at https://www.alexandervarwijk.com/blog/2020-11-18-graphql-in-the-open-social-drupal-distribution/

## Screenshots
afbeelding

### Issue 3174544 & 3174547
Open Social now includes an experimental GraphQL module. This module is in active development and no support will be provided for it at this time.

### Issue 3191701
User creation in tests no longer fails if no profile types have been configured.

### Issue 3191702
The schema for the social_user module has been fixed and true/false values are now proper booleans.

### Issue 3191703
It's now possible to attach libraries to account_header elements using the `#attached` key.

### Issue 3191704
The Open Social distribution now supports all Drupal PHPUnit test types and will execute them in our CI.

## Change Record
Omitted for now since we don't provide support for the GraphQL module. A change record should be created when we accept that distro users enabled the module.

## Translations
N.A.

### CODE REVIEW

File: modules/custom/social_graphql/src/Plugin/GraphQL/Schema/OpenSocialBaseSchema.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+   * build on.
+   *
+   * @param \Drupal\graphql\GraphQL\ResolverRegistryInterface $registry
+   *   The resolver registry.
+   */
+  protected function getBaseSchema(ResolverRegistryInterface $registry) {
+    $builder = new ResolverBuilder();
+
+    // TextFormat fields.
+    $registry->addFieldResolver('TextFormat', 'name',
```
Comment (ronaldtebrake) [inline modules/custom/social_graphql/src/Plugin/GraphQL/Schema/OpenSocialBaseSchema.php:60]: I think this is no longer valid :trollface:

Comment (Kingdutch) [reply modules/custom/social_graphql/src/Plugin/GraphQL/Schema/OpenSocialBaseSchema.php:65]: @ronaldtebrake Actually it is! We currently expose the text format for something to the client: https://github.com/goalgorilla/open_social/pull/1994/files#diff-0490ff176a1c7da94e6c07fc1f954fe341e8c0e003e4314ca2006ac2595bd877R167 although we may have to find a better way to provide clients with these details without having to be familiar with the meaning of our text formats.

**Comment by Kingdutch** [comment]
The following changes are included in this PR that will affect things even when `social_graphql` is disabled. All other changes will have no effect unless enabled for internal testing.

De volgende veranderingen zijn de enige veranderingen in de PR die buiten GraphQL plaatsvinden en dus invloed hebben als de module uit staat:

- Defensive [check for existance of profile type in social_profile.module](https://github.com/goalgorilla/open_social/pull/1994/files#diff-7a29143459605781e4114b267a1073edd03fbe24b18b88fae7baa5586e99ce1a)
- [Type fix in install/social_user.navigation.settings.yml](https://github.com/goalgorilla/open_social/pull/1994/files#diff-518763f38fa4fbcd8e489eaf726ff9ad456bf9835a28bb5a66b4b2fcc35e5029)
- [Schema fix for social_user module](https://github.com/goalgorilla/open_social/pull/1994/files#diff-d3097168bf7be5b7c485748ddbd7a62e5ea6bb2a0aca5e30e4f14ef67c765052) (schema werd nog niet toegepast)
- [Update hook fix for schema type in social_user.install](https://github.com/goalgorilla/open_social/pull/1994/files#diff-8cb80cb3bd580ce098336c97e55512f89090ee6846b3bfcf5352ed49c000fd6e)
- [`account_header` element now allows attaching libraries](https://github.com/goalgorilla/open_social/pull/1994/files#diff-8b5aebea7458bc55a71f22c95d5a752cf012a59b57714f622c20a138128763df)
- The unit test discovery has changed in [phpunit.xml](https://github.com/goalgorilla/open_social/pull/1994/files#diff-e35810879ec42bdd81797b3ccb72f6de28a8b4a0e3bfdba43183e133e866b892) and [.travis.yml](https://github.com/goalgorilla/open_social/pull/1994/files#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485). These changes are tested by Travis

**Comment by ronaldtebrake** [comment]
Just discussed with @Kingdutch if we can manage to ensure everything turns green we can add this to 10.0.0
We've marked it as hidden and explained Open Social now includes this as an experimental module. It's in active development and no support will be provided for it at this time.

The included updates outside of social_graphql can be cherry-picked where necessary. 

So the impact is low but gives us an easier roll-out plan for when it's ready to be deployed.

File: composer.json [hunk 2]
```diff
@@ -112,6 +113,9 @@
             },
             "drupal/views_infinite_scroll" : {
                 "Headers in table format repeat on load more instead of adding rows": "https://www.drupal.org/files/issues/2019-08-15/table_tbody_append-2899705-26.patch"
+            },
+            "webonyx/graphql-php": {
```
Comment (ronaldtebrake) [inline composer.json:117]: This does require ext-mbstring, which Drupal doesn't. Drupal uses symfony polyfill for this. Not sure if that causes any harm 🤔 could potentialy cause some issues for existing installions I guess 🤔

Comment (ronaldtebrake) [reply composer.json:117]: https://www.drupal.org/docs/system-requirements/php-requirements#s-mbstring-

Drupal says it does require it, also if it's not installed it seems to break https://www.drupal.org/node/2560123
So it should already be used everywhere.

**Pull request merged**